cvmfs/revision.py

```diff
@@ -0,0 +1,226 @@
+    represents the concrete status of the repository in a certain period of
+    time. Revision data is contained in the so-called Tags, which are stored in
+    the History database.
+    """
+
+    def __init__(self, repository, tag):
+        self.repository = repository
+        self._tag = tag
+
+    def __str__(self):
+        return '<Revision ' + self.revision_number \
+               + ' - ' + self.root_hash + '>'
+
+    @property
+    def revision_number(self):
+        return self._tag.revision
+
+    @property
+    def name(self):
+        return self._tag.name
+
+    @property
+    def timestamp(self):
+        return self._tag.timestamp
+
+    @property
+    def root_hash(self):
+        return self._tag.hash
+
+    def retrieve_catalog(self, catalog_hash):
+        """ Download and open a catalog that belongs to this revision """
+        return self.repository.retrieve_catalog(catalog_hash)
+
+    def retrieve_root_catalog(self):
+        return self.retrieve_catalog(self.root_hash)
+
+    def catalogs(self):
+        return CatalogTreeIterator(self)
+
+    def retrieve_catalog_for_path(self, needle_path):
+        """
+        Recursively walk down the Catalogs and find the best fit for a path
+        """
+        clg = self.retrieve_root_catalog()
+        while True:
+            new_nested_reference = clg.find_nested_for_path(needle_path)
+            if new_nested_reference is None:
+                break
+            nested_reference = new_nested_reference
+            clg = self.retrieve_catalog(nested_reference.hash)
+        return clg
+
+    def _opened_catalogs_for_path(self, path):
+        """
+        Gets the closest already opened catalog for a given path
+        :param path: the path to search for
+        :return: the closest opened catalog for a given path
+        """
+        best_catalog = self.retrieve_root_catalog()
+        max_length = 0
+        for catalog in self.repository.opened_catalogs.values():
+            curr_length = len(catalog.root_prefix) \
+                if path.find(catalog.root_prefix) == 0 else 0
+            if curr_length > max_length:
+                best_catalog = catalog
+                max_length = len(best_catalog.root_prefix)
+        return best_catalog
+
+    def _lookup_path(self, path):
+        """
+        Lookups in all existing catalogs for this path's best fit
+        :param path: path to search for
+        :return: the DirectoryEntry that corresponds to the given path if
+        it is found in the already loaded catalogs, or None otherwise
+        """
+        if path == '/':
+            path = ''
+        best_fit = self._opened_catalogs_for_path(path)
+        result = best_fit.find_directory_entry(path)
+        while result is None:
+            best_nested = best_fit.find_best_child_for_path(path)
+            if best_nested is None:
+                break
+            best_fit = best_nested.retrieve_from(self)
```

Comment:
Shouldn't that be retrieving the new nested catalog from the `repository`? Probably both works, because `Revision` implements a `retrieve_catalog()` method but I feel it would make more sense to delegate that to the corresponding `Repository` object.